src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -1,18 +1,654 @@
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.RobotMap;
+import frc.robot.utils.GeometryUtils;
+import frc.robot.utils.MatchStateUtil;
+import frc.robot.utils.PoseBuffer;
+import java.util.List;
+import java.util.Optional;
 
 public class DriveSubsystem extends SubsystemBase {
-    /* TODO: THIS SUBSYSTEM IS A PLACEHOLDER AND ONLY CONTAINS PLACEHOLDER METHODS */
-    public void driveToPoint(Pose2d point){
-        /* PLACEHOLDER METHOD */
+
+  private double targetHeadingDegrees;
+
+  double tagTargetHeading = 0.0;
+
+  // Create SwerveModules
+  public final SwerveModule frontLeft = new SwerveModule(
+      RobotMap.FRONT_LEFT_DRIVING_CAN_ID,
+      RobotMap.FRONT_LEFT_TURNING_CAN_ID,
+      DriveCal.FRONT_LEFT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule frontRight = new SwerveModule(
+      RobotMap.FRONT_RIGHT_DRIVING_CAN_ID,
+      RobotMap.FRONT_RIGHT_TURNING_CAN_ID,
+      DriveCal.FRONT_RIGHT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule rearLeft = new SwerveModule(
+      RobotMap.REAR_LEFT_DRIVING_CAN_ID,
+      RobotMap.REAR_LEFT_TURNING_CAN_ID,
+      DriveCal.BACK_LEFT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule rearRight = new SwerveModule(
+      RobotMap.REAR_RIGHT_DRIVING_CAN_ID,
+      RobotMap.REAR_RIGHT_TURNING_CAN_ID,
+      DriveCal.REAR_RIGHT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  private final Pigeon2 gyro = new Pigeon2(RobotMap.PIGEON_CAN_ID);
+  private ChassisSpeeds lastSetChassisSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+  public Optional<Pose2d> targetPose = Optional.empty();
+
+  public PoseBuffer poseBuffer = new PoseBuffer();
+
+  double KeepHeadingPID = 0.0;
+  double KeepHeadingFF = 0.0;
+
+  SwerveDriveOdometry odometry = new SwerveDriveOdometry(
+      DriveConstants.DRIVE_KINEMATICS,
+      Rotation2d.fromDegrees(Constants.PLACEHOLDER_DOUBLE),
+      getModulePositions());
+
+  public double throttleMultiplier = 1.0;
+
+  private double rotControllerInput = 0.0;
+
+  public MatchStateUtil matchState;
+
+  private InterpolatingDoubleTreeMap yawOffsetMap;
+
+  private InterpolatingDoubleTreeMap velocityToMultiplierMap;
+
+  public DriveSubsystem(MatchStateUtil matchState) {
+    intializeGyro();
+    this.matchState = matchState;
+
+    RobotConfig config;
+    try {
+      config = RobotConfig.fromGUISettings();
+      AutoBuilder.configure(
+        this::getPose, // Robot pose supplier
+        this::resetOdometry, // Method to reset odometry (will be called if your auto has a starting
+        // pose)
+        this::getCurrentChassisSpeeds, // ChassisSpeeds supplier. MUST BE ROBOT RELATIVE
+        (speeds, feedforwards) -> setOutputRobotRelativeSpeeds(speeds), // Method that will drive the robot given ROBOT RELATIVE
+        // ChassisSpeeds
+        new PPHolonomicDriveController(DriveCal.PATH_TRANSLATION_CONTROLLER, DriveCal.PATH_ROTATION_CONTROLLER),
```

Comment:
don't you define a PPHolonomicDriveController in DriveCal and/or Constants? Why then define a new one here?